torch/jit/_backends.py

```diff
@@ -0,0 +1,63 @@
+"""Backend API
+This module contains helper functions and utilities for the JIT backend
+API.
+"""
+
+from torch.jit._script import (
+    RecursiveScriptModule,
+)
+from typing import Callable, List
+
+
+# Type for functions that lower RecursiveScriptModules to JIT backends.
+ToBackendFnTy = Callable[
+    [RecursiveScriptModule], RecursiveScriptModule
+]
+
+
+def selective_to_jit_backend(
+    module: RecursiveScriptModule,
+    to_backend_fn: ToBackendFnTy,
+    modules_to_lower: List[str],
+):
+    """
+    Selectively lower modules in a module hierarchy to a JIT backend.
+
+
+    Arguments:
+        module: The RecursiveScriptModule at the root of the module hierachy that
+                    is to be selectively lowered.
+        to_backend_fn:  Function that will be called on each submodule in modules_to_lower
+                            to lower it to a JIT backend.
+        modules_to_lower: List of fully qualified names of submodules to selectively
+                            lower.
+    """
+    base = ""
+    selective_to_jit_backend_impl(module, to_backend_fn, modules_to_lower, base)
+
+
+def selective_to_jit_backend_impl(
+    module: RecursiveScriptModule,
+    to_backend_fn: ToBackendFnTy,
+    modules_to_lower: List[str],
+    base: str,
+) -> RecursiveScriptModule:
+    # For each submodule:
+    for name, submodule in module._modules.items():
+        # This is the fully qualified "path" to the attribute from the root module.
+        sep = "." if base else ""
+        qual_name = f"{base}{sep}{name}"
+        if qual_name in modules_to_lower:
+            # Lower the submodule.
+            lowered_submodule = to_backend_fn(submodule)
+            module_jit_type = module._c._type()
+            # Modify the corresponding attribute on the owning module's type.
+            module_jit_type.unsafe_remove_attribute(name)
+            module_jit_type.add_attribute(name, lowered_submodule._c._type())
```

Comment:
One thing that worries me is that by default, types are shared across different calls to `torch.jit.script`. So if a user has multiple instances of `CoolModule`, scripts them all and selectively lowers a submodule of one of them, then this will modify the JIT type of all `CoolModule` instances. I see two options:

1) Let the user explicitly opt out of type sharing if they want to do this (probably a bad idea)
2) Try to mimic some sort of copy-on-write behaviour where we clone the type we want to edit just before we do this.

Reply:
I think it would be safer/easier to just copy the entire module hierarchy and perform the edits on that. That's the approach that quantization takes, I believe there is already cloning + typemapping logic hanging off `Module` somewhere.